File: packages/cli/src/migrations/repo-format-v1.test.ts
import {
  lstat,
  mkdir,
  mkdtemp,
  readFile,
  rm,
  writeFile,
} from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { AppConstants } from "#app/config/constants.ts";
import type { ResolvedSyncConfig } from "#app/config/sync-schema.ts";
import { createSymlink } from "#app/lib/filesystem.ts";
import { migrateRepositoryFormatV0ToV1 } from "./repo-format-v1.ts";

let repoDir: string;

const emptyConfig: ResolvedSyncConfig = {
  entries: [],
  profiles: [],
  version: 8,
};

const suffix = AppConstants.SYNC.SYMLINK_ARTIFACT_SUFFIX;

beforeEach(async () => {
  repoDir = await mkdtemp(join(tmpdir(), "dotweave-repo-format-"));
});

afterEach(async () => {
  await rm(repoDir, { force: true, recursive: true });
});

describe("migrateRepositoryFormatV0ToV1", () => {
  it("does nothing when there is no profiles directory", async () => {
    await expect(
      migrateRepositoryFormatV0ToV1(repoDir, emptyConfig),
    ).resolves.toBeUndefined();
  });

  it("converts physical symlink artifacts (nested) to metadata files and is idempotent", async () => {
    const profileDir = join(repoDir, "profiles", "default", ".config");
    await mkdir(profileDir, { recursive: true });

    // A regular file that must be left untouched.
    await writeFile(join(profileDir, "keep.txt"), "keep\n", "utf8");

    // A physical file symlink artifact.
    await createSymlink("../.agents/note.md", join(profileDir, "note.md"));

    // A physical directory symlink artifact one level deeper.
    const nested = join(profileDir, "nested");
    await mkdir(nested, { recursive: true });
    await createSymlink("../../.agents/skills", join(nested, "skills"));

    await migrateRepositoryFormatV0ToV1(repoDir, emptyConfig);

    // Physical links are gone; metadata files carry the POSIX target.
    await expect(lstat(join(profileDir, "note.md"))).rejects.toThrow();
    expect(await readFile(join(profileDir, `note.md${suffix}`), "utf8")).toBe(
      "../.agents/note.md",
    );

    await expect(lstat(join(nested, "skills"))).rejects.toThrow();
    expect(await readFile(join(nested, `skills${suffix}`), "utf8")).toBe(
      "../../.agents/skills",
    );

    // Regular file preserved.
    expect(await readFile(join(profileDir, "keep.txt"), "utf8")).toBe("keep\n");

    // Metadata files are regular files, not symlinks.
    expect(
      (await lstat(join(profileDir, `note.md${suffix}`))).isSymbolicLink(),
    ).toBe(false);

    // Idempotent: a second run makes no further changes and does not throw.
    await migrateRepositoryFormatV0ToV1(repoDir, emptyConfig);
    expect(await readFile(join(profileDir, `note.md${suffix}`), "utf8")).toBe(
      "../.agents/note.md",
    );
  });
});

File: packages/cli/test/migrations/repo_format_v1_test.dart
import 'dart:io';

import 'package:dotweave/src/config/constants.dart';
import 'package:dotweave/src/config/repo_format_migration.dart';
import 'package:dotweave/src/lib/filesystem.dart';
import 'package:dotweave/src/migrations/repo_format_v1.dart';
import 'package:path/path.dart' as p;
import 'package:test/test.dart';

void main() {
  late String repoDir;

  const emptyConfig = ResolvedSyncConfig(entries: [], profiles: [], version: 8);

  final suffix = AppConstants.sync.symlinkArtifactSuffix;

  setUp(() async {
    repoDir = (await Directory.systemTemp.createTemp(
      'dotweave-repo-format-',
    )).path;
  });

  tearDown(() async {
    await removePath(repoDir);
  });

  group('migrateRepositoryFormatV0ToV1', () {
    test('does nothing when there is no profiles directory', () async {
      await expectLater(
        migrateRepositoryFormatV0ToV1(repoDir, emptyConfig),
        completes,
      );
    });

    test('converts physical symlink artifacts (nested) to metadata files '
        'and is idempotent', () async {
      final profileDir = p.join(repoDir, 'profiles', 'default', '.config');
      await Directory(profileDir).create(recursive: true);

      // A regular file that must be left untouched.
      await File(p.join(profileDir, 'keep.txt')).writeAsString('keep\n');

      // A physical file symlink artifact.
      await createSymlink('../.agents/note.md', p.join(profileDir, 'note.md'));

      // A physical directory symlink artifact one level deeper.
      final nested = p.join(profileDir, 'nested');
      await Directory(nested).create(recursive: true);
      await createSymlink('../../.agents/skills', p.join(nested, 'skills'));

      await migrateRepositoryFormatV0ToV1(repoDir, emptyConfig);

      // Physical links are gone; metadata files carry the POSIX target.
      expect(await getPathStats(p.join(profileDir, 'note.md')), isNull);
      expect(
        await File(p.join(profileDir, 'note.md$suffix')).readAsString(),
        '../.agents/note.md',
      );

      expect(await getPathStats(p.join(nested, 'skills')), isNull);
      expect(
        await File(p.join(nested, 'skills$suffix')).readAsString(),
        '../../.agents/skills',
      );

      // Regular file preserved.
      expect(
        await File(p.join(profileDir, 'keep.txt')).readAsString(),
        'keep\n',
      );

      // Metadata files are regular files, not symlinks.
      expect(
        (await getPathStats(
          p.join(profileDir, 'note.md$suffix'),
        ))!.isSymbolicLink,
        isFalse,
      );

      // Idempotent: a second run makes no further changes and does not throw.
      await migrateRepositoryFormatV0ToV1(repoDir, emptyConfig);
      expect(
        await File(p.join(profileDir, 'note.md$suffix')).readAsString(),
        '../.agents/note.md',
      );
    });
  });
}
